src/lib/core/ssl.h

```diff
@@ -0,0 +1,29 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2021, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include "trivia/config.h"
+
+#if defined(ENABLE_SSL)
+# include "ssl_impl.h"
```

Comment:
EE's file is the same. Might be better to drop `ssl_impl.h` and its inclusions.

Reply:
You're right, but I'd rather follow the same procedure for all EE headers, for consistency. In the long run, we might want to add an inline function to a EE header, in which case we'd have to add `_impl.h` anyway so we'd better do it right way IMO. I documented the procedure [here](https://github.com/tarantool/tarantool-ee/blob/master/README.md#how-to-contribute).

Reply:
From the procedure I understood that you do it to never include `_impl.h` except in the OS `.c` file with the same name.
 - But it still would be possible if the only header file is the one in OS. If EE would have it exactly the same anyway.
 - You have to declare structs and classes both in the normal header and in the `_impl.h` file. How does it work that C++ compiler allows to define `class SSLError` twice? I didn't try to compile EE though.
 - You are talking about potential need to add a new function to the `_impl.h` file which apparently wouldn't be in the OS file. But that would mean you need to include that `_impl.h` file somewhere to use that function. Which defeats the purpose of `_impl.h` files - "never include them except in their OS-version `.c` file. If you would want to use that inline function inside of the impl file, it could also be done in the OS version. Or could become not inline.

Reply:
`XXX_impl.h` is only included by OS `XXX.h` in case it's a EE build. The rest of the content of `XXX.h` is ignored for EE builds. That's why `class SSLError` isn't defined twice:

https://github.com/tarantool/tarantool/blob/36bf0cb36077fefbee7fd638a398d78dadad2e69/src/lib/core/ssl_error.h#L10-L12

If we ever patch performance-sensitive parts of Tarantool we might want to make a EE function inline. The OS version of the same function may throw an error or just be unreachable. Hence we need a separate header for EE builds which would be included (directly or indirectly) by OS files. I think it's better to separate OS and EE headers completely, because it's a straightforward procedure (you don't need to think about which functions go to which header: OS header contains stubs while EE header does the implementation) and in future it'll allow us to easily open sources of EE features if we decide to (by simply moving EE's `XXX_impl.h` to OS `XXX.h`).

Reply:
Ok, as you wish. I only don't like code duplication. Because will need to patch 2 places. For instance, when errors are going to become just strings/enums instead of C++ classes. Anyway, if it works now, then ok. Can deal with it later if it will get too complicated.